src/gurt/telemetry.c

```diff
@@ -68,65 +99,356 @@ struct d_tm_node_t *
 d_tm_get_root(struct d_tm_context *ctx)
 {
 	if (ctx != NULL && ctx->shmem_root != NULL)
-		return d_tm_conv_ptr(ctx, ctx->shmem_root->sh_root);
+		return conv_ptr(ctx->shmem_root,
+				ctx->shmem_root->sh_root);
 
 	return NULL;
 }
 
 static int
 d_tm_lock_shmem(void)
 {
-	return D_MUTEX_LOCK(&tm_shmem.dts_add_lock);
+	return D_MUTEX_LOCK(&tm_shmem.add_lock);
 }
 
 static int
 d_tm_unlock_shmem(void)
 {
-	return D_MUTEX_UNLOCK(&tm_shmem.dts_add_lock);
+	return D_MUTEX_UNLOCK(&tm_shmem.add_lock);
+}
+
+/*
+ * Returns either the shmid or an error.
+ */
+static int
+attach_shmem(key_t key, size_t size, int flags, struct d_tm_shmem_hdr **shmem)
+{
+	int	 shmid;
+	void	*addr;
+
+	D_ASSERT(shmem != NULL);
+
+	shmid = shmget(key, size, flags);
+	if (shmid < 0) {
+		D_ERROR("can't get shmid for key 0x%x, %s\n", key,
+			strerror(errno));
+		return -DER_NO_SHMEM;
+	}
+
+	addr = shmat(shmid, NULL, 0);
+	if (addr == (void *)-1) {
+		D_ERROR("can't attach to shared memory, %s\n", strerror(errno));
+		return -DER_SHMEM_PERMS;
+	}
+
+	*shmem = addr;
+	return shmid;
+}
+
+static int
+new_shmem(key_t key, size_t size, struct d_tm_shmem_hdr **shmem)
+{
+	D_INFO("creating new shared memory segment, key=0x%x, size=%lu\n",
+	       key, size);
+	return attach_shmem(key, size, IPC_CREAT | 0660, shmem);
+}
+
+static int
+open_shmem(key_t key, struct d_tm_shmem_hdr **shmem)
+{
+	return attach_shmem(key, 0, 0, shmem);
+}
+
+inline static void
+close_shmem(struct d_tm_shmem_hdr *shmem)
+{
+	if (shmem == NULL)
+		return;
+
+	if (shmdt(shmem) < 0)
+		D_INFO("failed to detach from shmem segment, %s\n",
+			strerror(errno));
+}
+
+inline static void
```

Comment:
(style)  storage class should be at the beginning of the declaration
(style)  inline keyword should sit between storage class and type